File: client/javascript/src/mcp_bridge.js
/**
 * @file mcp_bridge.js
 * Local bridge from this Node.js client to MCP servers.
 *
 * Mirrors the Rust/Python clients' bridges: spawns child processes for
 * `stdio` transport, multiplexes JSON-RPC requests over a single
 * connection, and performs the MCP `initialize` handshake at registration
 * time so the engine-side registration is only announced if the server
 * actually came up.
 *
 * Browser environments don't have `child_process`; importing this module
 * is fine, but `BridgeRegistry.registerStdio` will throw at call time.
 */

const PROTOCOL_VERSION = '2024-11-05';
const CLIENT_INFO = { name: 'pie-client-js', version: '1.0.0' };

/**
 * A JSON-RPC error returned by an MCP server (or synthesized for
 * transport-level failures).
 */
export class JsonRpcError extends Error {
    constructor(code, message, data = null) {
        super(`JSON-RPC error ${code}: ${message}`);
        this.code = code;
        this.data = data;
    }
}

/**
 * A connection to a single stdio MCP server. Many caller flows
 * multiplex over the same connection using distinct request ids.
 *
 * Construct via `StdioServer.spawn(...)` rather than `new`.
 */
export class StdioServer {
    constructor(name, child) {
        this.name = name;
        this.child = child;
        this._nextId = 1;
        this._pending = new Map();
        this._dead = false;
        this._stdoutBuf = '';

        child.stdout.setEncoding('utf8');
        child.stdout.on('data', (chunk) => this._onStdoutData(chunk));
        child.stdout.on('end', () => this._markDead());

        child.stderr.setEncoding('utf8');
        child.stderr.on('data', (chunk) => {
            for (const line of String(chunk).split('\n')) {
                if (line) console.error(`[mcp:${this.name}] ${line}`);
            }
        });

        child.on('exit', () => this._markDead());
        child.on('error', () => this._markDead());
    }

    /**
     * Spawn an MCP server process. Resolves once the child is alive;
     * the caller is expected to call `handshake()` next.
     */
    static async spawn(name, command, args = []) {
        const { spawn } = await import('node:child_process');
        const child = spawn(command, args, {
            stdio: ['pipe', 'pipe', 'pipe'],
        });
        return new StdioServer(name, child);
    }

    async handshake() {
        await this.call('initialize', {
            protocolVersion: PROTOCOL_VERSION,
            capabilities: {},
            clientInfo: CLIENT_INFO,
        });
        // Notification — no response expected.
        this._writeJson({
            jsonrpc: '2.0',
            method: 'notifications/initialized',
            params: {},
        });
    }

    /**
     * Send a JSON-RPC request and await the matching response.
     * Resolves with the `result` field on success; rejects with
     * `JsonRpcError` on JSON-RPC `error` or transport failure.
     */
    call(method, params) {
        if (this._dead) {
            return Promise.reject(new JsonRpcError(
                -32000,
                `MCP server '${this.name}' is no longer running`,
            ));
        }
        return new Promise((resolve, reject) => {
            const id = this._nextId++;
            this._pending.set(id, { resolve, reject });
            try {
                this._writeJson({ jsonrpc: '2.0', id, method, params });
            } catch (err) {
                this._pending.delete(id);
                reject(new JsonRpcError(
                    -32000,
                    `MCP server '${this.name}' input closed: ${err.message}`,
                ));
            }
        });
    }

    /** @private */
    _writeJson(obj) {
        this.child.stdin.write(JSON.stringify(obj) + '\n');
    }

    /** @private */
    _onStdoutData(chunk) {
        this._stdoutBuf += chunk;
        let nl;
        while ((nl = this._stdoutBuf.indexOf('\n')) !== -1) {
            const line = this._stdoutBuf.slice(0, nl).trim();
            this._stdoutBuf = this._stdoutBuf.slice(nl + 1);
            if (!line) continue;
            this._handleLine(line);
        }
    }

    /** @private */
    _handleLine(line) {
        let msg;
        try {
            msg = JSON.parse(line);
        } catch (e) {
            console.error(`[mcp:${this.name}] non-JSON line (${e.message}): ${line}`);
            return;
        }
        const id = msg.id;
        if (typeof id !== 'number') return; // ignore notifications/server-initiated
        const slot = this._pending.get(id);
        if (!slot) return;
        this._pending.delete(id);
        if (msg.error) {
            const { code = -32000, message = 'MCP error', data = null } = msg.error;
            slot.reject(new JsonRpcError(code, message, data));
        } else {
            slot.resolve(msg.result ?? null);
        }
    }

    /** @private */
    _markDead() {
        if (this._dead) return;
        this._dead = true;
        for (const slot of this._pending.values()) {
            slot.reject(new JsonRpcError(
                -32000,
                `MCP server '${this.name}' died before response`,
            ));
        }
        this._pending.clear();
    }

    /**
     * Graceful shutdown ladder: close stdin (most MCP servers exit on
     * EOF) → SIGTERM → SIGKILL. Each step gives the server a brief
     * window to exit before escalating.
     */
    async close() {
        if (this.child.exitCode !== null || this.child.signalCode !== null) return;
        try { this.child.stdin.end(); } catch {}
        if (await this._waitForExit(1000)) return;
        try { this.child.kill('SIGTERM'); } catch {}
        if (await this._waitForExit(1000)) return;
        try { this.child.kill('SIGKILL'); } catch {}
        await this._waitForExit(1000);
    }

    /** @private */
    _waitForExit(ms) {
        return new Promise((resolve) => {
            if (this.child.exitCode !== null || this.child.signalCode !== null) {
                resolve(true);
                return;
            }
            const onExit = () => { clearTimeout(t); resolve(true); };
            const t = setTimeout(() => {
                this.child.removeListener('exit', onExit);
                resolve(false);
            }, ms);
            this.child.once('exit', onExit);
        });
    }
}

/**
 * Per-`PieClient` registry of locally-spawned MCP servers.
 */
export class BridgeRegistry {
    constructor() {
        this._servers = new Map();
    }

    /**
     * Spawn a stdio MCP server, complete the `initialize` handshake,
     * and publish it under `name`. Throws if `name` is already
     * registered or the handshake fails.
     */
    async registerStdio(name, command, args = []) {
        if (this._servers.has(name)) {
            throw new Error(`MCP server '${name}' already registered`);
        }
        const server = await StdioServer.spawn(name, command, args);
        try {
            await server.handshake();
        } catch (err) {
            await server.close();
            throw err;
        }
        this._servers.set(name, server);
    }

    get(name) {
        return this._servers.get(name) ?? null;
    }

    async closeAll() {
        const all = Array.from(this._servers.values());
        this._servers.clear();
        await Promise.all(all.map((s) => s.close()));
    }
}

File: client/rust/src/mcp_bridge.rs
//! Local bridge from this client to MCP servers.
//!
//! When the engine asks us to relay an MCP call (because some inferlet called
//! `mcp::client::session::call_tool` etc.), we look up the named server here
//! and forward the call as JSON-RPC over the server's transport.
//!
//! Currently only the `stdio` transport is implemented: we spawn a child
//! process, write line-delimited JSON-RPC requests to its stdin, and read
//! line-delimited JSON-RPC responses from its stdout. One child per
//! registered name; many inferlets multiplex over the single connection
//! using distinct request ids.

use anyhow::{Context, Result, anyhow, bail};
use dashmap::DashMap;
use serde_json::{Value, json};
use std::process::Stdio;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicI64, Ordering};
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::process::Command;
use tokio::sync::oneshot;
use tokio::task::JoinHandle;

const PROTOCOL_VERSION: &str = "2024-11-05";

/// JSON-RPC error decoded from an MCP server response.
#[derive(Debug, Clone)]
pub struct JsonRpcError {
    pub code: i64,
    pub message: String,
    pub data: Option<Value>,
}

impl std::fmt::Display for JsonRpcError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "JSON-RPC error {}: {}", self.code, self.message)
    }
}

impl std::error::Error for JsonRpcError {}

/// Per-`Client` registry of locally-spawned MCP servers.
pub struct BridgeRegistry {
    servers: DashMap<String, Arc<StdioServer>>,
}

impl std::fmt::Debug for BridgeRegistry {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let names: Vec<String> = self.servers.iter().map(|r| r.key().clone()).collect();
        f.debug_struct("BridgeRegistry").field("servers", &names).finish()
    }
}

impl BridgeRegistry {
    pub fn new() -> Self {
        Self { servers: DashMap::new() }
    }

    /// Spawn a stdio MCP server, complete the `initialize` handshake, and
    /// publish it under `name`. Fails if `name` is already registered or if
    /// the handshake doesn't succeed.
    pub async fn register_stdio(
        &self,
        name: &str,
        command: &str,
        args: &[String],
    ) -> Result<()> {
        if self.servers.contains_key(name) {
            bail!("MCP server '{}' already registered", name);
        }
        let server = StdioServer::spawn(name, command, args).await?;
        server.handshake().await
            .with_context(|| format!("MCP handshake with '{}' failed", name))?;
        self.servers.insert(name.to_string(), Arc::new(server));
        Ok(())
    }

    pub fn get(&self, name: &str) -> Option<Arc<StdioServer>> {
        self.servers.get(name).map(|r| Arc::clone(r.value()))
    }
}

impl Default for BridgeRegistry {
    fn default() -> Self { Self::new() }
}

// =============================================================================
// StdioServer
// =============================================================================

/// State shared between an MCP server's I/O tasks and the public handle.
struct StdioInner {
    name: String,
    stdin_tx: tokio::sync::mpsc::UnboundedSender<String>,
    pending: DashMap<i64, oneshot::Sender<std::result::Result<Value, JsonRpcError>>>,
    next_id: AtomicI64,
    dead: AtomicBool,
}

/// Aborts a `JoinHandle` when dropped. Used to tie task lifetimes to the
/// outer handle so that dropping the registry entry tears down the child
/// without leaking tasks.
struct AbortOnDrop(JoinHandle<()>);
impl Drop for AbortOnDrop { fn drop(&mut self) { self.0.abort(); } }

/// A connection to a single stdio MCP server. Cheap to clone (`Arc`), all
/// methods route through the multiplexed JSON-RPC channel.
pub struct StdioServer {
    inner: Arc<StdioInner>,
    // Field order matters: Rust drops fields top-to-bottom, so on Drop we
    // first abort the supervisor (whose `Child` is `kill_on_drop` → process
    // dies), which lets reader/writer end naturally on EOF.
    _supervisor: AbortOnDrop,
    _reader: AbortOnDrop,
    _writer: AbortOnDrop,
    _stderr: AbortOnDrop,
}

impl StdioServer {
    async fn spawn(name: &str, command: &str, args: &[String]) -> Result<Self> {
        let mut child = Command::new(command)
            .args(args)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true)
            .spawn()
            .with_context(|| format!("Failed to spawn MCP server '{}': {}", name, command))?;

        let stdin = child.stdin.take().context("MCP child has no stdin")?;
        let stdout = child.stdout.take().context("MCP child has no stdout")?;
        let stderr = child.stderr.take().context("MCP child has no stderr")?;

        let (line_tx, mut line_rx) = tokio::sync::mpsc::unbounded_channel::<String>();
        let inner = Arc::new(StdioInner {
            name: name.to_string(),
            stdin_tx: line_tx,
            pending: DashMap::new(),
            next_id: AtomicI64::new(1),
            dead: AtomicBool::new(false),
        });

        // stdin writer
        let writer = tokio::spawn(async move {
            let mut stdin = stdin;
            while let Some(line) = line_rx.recv().await {
                if stdin.write_all(line.as_bytes()).await.is_err() { break; }
                if stdin.write_all(b"\n").await.is_err() { break; }
                if stdin.flush().await.is_err() { break; }
            }
        });

        // stdout reader — routes JSON-RPC responses to pending callers
        let reader_inner = Arc::clone(&inner);
        let reader = tokio::spawn(async move {
            let mut br = BufReader::new(stdout);
            let mut line = String::new();
            loop {
                line.clear();
                match br.read_line(&mut line).await {
                    Ok(0) | Err(_) => break,
                    Ok(_) => {}
                }
                let trimmed = line.trim();
                if trimmed.is_empty() { continue; }
                let v: Value = match serde_json::from_str(trimmed) {
                    Ok(v) => v,
                    Err(e) => {
                        eprintln!("[mcp:{}] non-JSON line ({}): {}", reader_inner.name, e, trimmed);
                        continue;
                    }
                };
                // We only handle responses (id present + result|error).
                // Server-initiated requests/notifications are ignored for now.
                let id = match v.get("id").and_then(Value::as_i64) {
                    Some(i) => i,
                    None => continue,
                };
                if let Some((_, sender)) = reader_inner.pending.remove(&id) {
                    let outcome = if let Some(err) = v.get("error") {
                        let code = err.get("code").and_then(Value::as_i64).unwrap_or(-32000);
                        let message = err.get("message").and_then(Value::as_str)
                            .unwrap_or("MCP error").to_string();
                        let data = err.get("data").cloned();
                        Err(JsonRpcError { code, message, data })
                    } else {
                        Ok(v.get("result").cloned().unwrap_or(Value::Null))
                    };
                    let _ = sender.send(outcome);
                } else {
                    eprintln!("[mcp:{}] response for unknown id {}", reader_inner.name, id);
                }
            }
            // EOF / error: mark dead and fail any outstanding pendings.
            reader_inner.dead.store(true, Ordering::SeqCst);
            reader_inner.pending.clear();
        });

        // stderr drain — surface server errors to the client's stderr
        let name_for_err = name.to_string();
        let stderr_task = tokio::spawn(async move {
            let mut br = BufReader::new(stderr);
            let mut line = String::new();
            loop {
                line.clear();
                match br.read_line(&mut line).await {
                    Ok(0) | Err(_) => break,
                    Ok(_) => {}
                }
                eprint!("[mcp:{}] {}", name_for_err, line);
            }
        });

        // Supervisor — owns Child, reaps on exit. kill_on_drop ensures the
        // process dies if the supervisor task is aborted.
        let supervisor = tokio::spawn(async move {
            let _ = child.wait().await;
        });

        Ok(Self {
            inner,
            _supervisor: AbortOnDrop(supervisor),
            _reader: AbortOnDrop(reader),
            _writer: AbortOnDrop(writer),
            _stderr: AbortOnDrop(stderr_task),
        })
    }

    /// Perform the MCP `initialize` handshake and send `initialized`.
    async fn handshake(&self) -> Result<()> {
        let init_params = json!({
            "protocolVersion": PROTOCOL_VERSION,
            "capabilities": {},
            "clientInfo": {
                "name": "pie-client",
                "version": env!("CARGO_PKG_VERSION"),
            },
        });
        self.call("initialize", init_params).await
            .map_err(|e| anyhow!("{}", e))?;
        // Notification — no response expected.
        self.notify("notifications/initialized", json!({}))?;
        Ok(())
    }

    /// Send a JSON-RPC request and await the matching response.
    pub async fn call(
        &self,
        method: &str,
        params: Value,
    ) -> std::result::Result<Value, JsonRpcError> {
        if self.inner.dead.load(Ordering::SeqCst) {
            return Err(JsonRpcError {
                code: -32000,
                message: format!("MCP server '{}' is no longer running", self.inner.name),
                data: None,
            });
        }
        let id = self.inner.next_id.fetch_add(1, Ordering::SeqCst);
        let (tx, rx) = oneshot::channel();
        self.inner.pending.insert(id, tx);
        let req = json!({
            "jsonrpc": "2.0",
            "id": id,
            "method": method,
            "params": params,
        });
        if self.inner.stdin_tx.send(req.to_string()).is_err() {
            self.inner.pending.remove(&id);
            return Err(JsonRpcError {
                code: -32000,
                message: format!("MCP server '{}' input closed", self.inner.name),
                data: None,
            });
        }
        match rx.await {
            Ok(outcome) => outcome,
            Err(_) => {
                self.inner.pending.remove(&id);
                Err(JsonRpcError {
                    code: -32000,
                    message: format!("MCP server '{}' died before response", self.inner.name),
                    data: None,
                })
            }
        }
    }

    /// Send a JSON-RPC notification (no id, no response).
    fn notify(&self, method: &str, params: Value) -> Result<()> {
        let req = json!({
            "jsonrpc": "2.0",
            "method": method,
            "params": params,
        });
        self.inner.stdin_tx.send(req.to_string())
            .map_err(|_| anyhow!("MCP server '{}' input closed", self.inner.name))?;
        Ok(())
    }
}
